### PR TITLE
Allow GC on the first instruction of NoGC region (except in prologs/epilogs)

### PR DESCRIPTION
**Pull request by VSadov**
NoGC regions are used to temporarily disable GC in otherwise interruptible code. A typical use is during block-copying data that may contain GC references via temporaries that are not GC-reported.

The semantic of NoGC region is that once the first instruction executes, tracking invariants are violated and GC cannot happen, until the execution of the last instruction in the region makes GC safe again.  
In particular - once the IP is on the first instruction, but not executed it yet, it is still safe to do GC. 

The only special case is when NoGC region is used for prologs/epilogs. In such case the GC info could be incorrect until the prolog completes and epilogs may have unwindability restrictions, so the first instruction cannot have GC.

======   What prompted this change:

While working on https://github.com/dotnet/runtime/pull/102680. I have discovered that sometimes we start NoGC region on the next instruction after a call. Without this change it means that once the call executes, the caller is in a NoGC state. 
As a result there is a rare combination where:
- the calee was hijacked and upon return tries to initiate a stackwalk at the call site in the caller. 
- the caller is fully-interruptible, thus it checks if a stackwalk starts in interruptible code.

That leads to an assert in `gcinfodecoder.cpp:801`  `executionAborted`.
We do not expect to be asked for GC info in non-interruptible code unless it is the `executionAborted` case.

It would be nice to not mark instructions as NoGC unnecessarily. In particular we should not do that to GC-capable call sites.

### CODE REVIEW

**Comment by VSadov** [comment]
The stress failures look preexisting, but the failure on ` Libraries Test Run checked coreclr osx x64 Debug` is new. Not sure what is special about that configuration.

**Comment by VSadov** [comment]
As compared with a baseline run https://github.com/dotnet/runtime/pull/85694 - the stress failures appear the same or fewer.

**Comment by VSadov** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/codegenarm.cpp [hunk 1]
```diff
@@ -119,12 +119,12 @@ BasicBlock* CodeGen::genCallFinally(BasicBlock* block)
 {
     assert(block->KindIs(BBJ_CALLFINALLY));
 
-    GetEmitter()->emitIns_J(INS_bl, block->GetTarget());
-
     BasicBlock* nextBlock = block->Next();
 
     if (block->HasFlag(BBF_RETLESS_CALL))
     {
+        GetEmitter()->emitIns_J(INS_bl, block->GetTarget());
```
Comment (VSadov) [inline src/coreclr/jit/codegenarm.cpp:126]: Unlike other uses of NoGC regions the use in `genCallFinally` was starting the NoGC region on instruction that already cannot do GC. 

We need to agree on the same semantics for all the uses, so `genCallFinally` had to be adjusted to work the same as in other uses of NoGC like block copying, tailcall arg setups,...

Comment (VSadov) [reply src/coreclr/jit/codegenarm.cpp:126]: We could have it this way (and adjust uses in block copying, etc..), we just need to have it the same way in all cases. 

However, starting NoGC region on an instruction group that _will_ make GC unsafe seems a lot more convenient as we may want to start NoGC region before we know exact instructions that will go into it.

File: src/coreclr/jit/gcencode.cpp [hunk 2]
```diff
@@ -4013,37 +4013,43 @@ void GCInfo::gcInfoBlockHdrSave(GcInfoEncoder* gcInfoEncoder, unsigned methodSiz
...
+        if (igOffs > m_uninterruptibleEnd)
         {
-            gcInfoEncoderWithLog->DefineInterruptibleRange(prevStart, igOffs - prevStart);
+            // Once the first instruction in IG executes, we cannot have GC.
+            // But it is ok to have GC while the IP is on the first instruction, unless we are in prolog/epilog.
+            unsigned interruptibleEnd = igOffs;
+            if (!isInPrologOrEpilog)
+            {
+                interruptibleEnd += firstInstrSize;
+            }
```
Comment (VSadov) [inline src/coreclr/jit/gcencode.cpp:4049]: This is the key change in this PR.

File: src/coreclr/jit/gcencode.cpp [hunk 3]
```diff
@@ -4013,37 +4013,43 @@ void GCInfo::gcInfoBlockHdrSave(GcInfoEncoder* gcInfoEncoder, unsigned methodSiz
 //
 struct InterruptibleRangeReporter
 {
-    unsigned prevStart;
-    Encoder* gcInfoEncoderWithLog;
+    unsigned m_uninterruptibleEnd;
+    Encoder* m_gcInfoEncoder;
```
Comment (jakobbotsch) [inline src/coreclr/jit/gcencode.cpp:4017]: Make them private as well?

**Review by jakobbotsch — APPROVED** [review]
LGTM. Can you add some documentation on `emitDisableGC` (using [new-style function header](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#94-function-header-comment)) that GC will be disabled only once the next instruction emitted has been executed?

**Comment by VSadov** [comment]
> Can you add some documentation on emitDisableGC (using [new-style function header](https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/clr-jit-coding-conventions.md#94-function-header-comment)) that GC will be disabled only once the next instruction emitted has been executed?

Good point, Added.

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**